### PR TITLE
[1단계 - 지하철 정보 관리 기능] 다즐(최우창) 미션 제출합니다.

### PR DESCRIPTION
**Pull request by woo-chang**
안녕하세요 토니 🙇🏻‍♂️ 다즐입니다! 블랙잭 이후 다시 만났네요 그동안 잘 지내셨나요?

이번 지하철 미션에서는 페어로 포이를 만나 많은 얘기를 나누며 재밌게 페어 프로그래밍을 해보았습니다. 이번 미션에서 핵심으로 가져가고자 했던 부분은 **도메인**과 **테스트**입니다. 그동안 웹 애플리케이션 개발을 진행할 때는 먼저 DB 설계를 진행한 뒤, 엔티티를 도메인으로 보고 DB 중심적인 개발을 했었는데 이번 미션에서는 도메인을 먼저 설계, 구현하고 DB를 가장 마지막에 적용해 보았습니다. 웹 애플리케이션은 DB 설계가 우선되어야 한다는 고정 관념이 있었는데 이번 미션을 통해 고정 관념이 깨져버리게 되었습니다.

또한 TDD를 지키고자 노력하였고 controller -> service -> repository 순으로 테스트를 진행해 보았습니다. 하위 계층에 대한 구현을 mocking 하여 해당 계층에서의 역할을 중심적으로 테스트하려고 했던 것 같아요. 모든 기능 구현 후에는 통합 테스트도 진행하여 애플리케이션 전체 동작도 테스트해보았습니다. 컨트롤러에서 서비스 계층에 대해 mocking 했던 코드만 삭제하였더니 통합 테스트도 정상적으로 수행하였습니다.

추가로 미션을 진행하면서 생긴 질문에 대해 아래에 정리해보았습니다!

## DB에 있는 값은 안전하다고 판단할 수 있을까?

도메인을 중심으로 개발을 진행하였기에 노선에 역을 추가하거나, 제거하기 위해서는 DB에 있는 데이터를 바탕으로 노선이라는 도메인을 만들어 작업하게 됩니다. 여기서 **DB에 저장된 값은 도메인에게 있어 항상 유효한 값일 수 있는가**에 대한 의문이 들었습니다. 물론 저장과정에도 도메인을 바탕으로 데이터가 저장되고 조회 시에도 도메인으로 다시 변환하기에 검증은 되지만 혹시 모를 유효하지 않은 값이 있을까에 대한 궁금증이 남아있습니다. DB의 값은 항상 유효한 데이터라고 판단해도 괜찮은 걸까요? 🤔

## 도메인 로직

도메인에 로직이 들어가기에 service 계층에서는 SQL로 비즈니스 로직을 처리하기보다는 도메인의 상태를 가지고 비즈니스 로직을 처리하게 되었습니다. 역이 연결된 노선에 역을 추가하는 상황을 예로 들면 아래와 같은 고민을 해볼 수 있을 것 같아요.

스크린샷 2023-05-11 오후 4 44 09

쿼리를 사용한다면 노선을 온전히 알지 않아도 풀 수 있는 로직인데 노선이라는 도메인이 들어감으로 노선을 온전히 생성해서 사용이 필요했습니다. 한편으로는 노선 도메인을 통해 복잡한 쿼리를 날리지 않아도 되는 장점이 존재했습니다. 두 방법 다 장단점이 있는 것 같아서 어떤 상황에 어떤 선택을 해야 할지 아직 감이 오지 않는 것 같습니다. 도메인을 사용하더라도 필요에 따라 쿼리를 통해 문제 상황을 풀어가는 경우도 존재하나요? 🤔

그럼에도 도메인을 사용하는 이유는 성능상의 이점을 조금 포기하더라도 애플리케이션을 **복잡성을 낮춰 유지보수하기 좋은 애플리케이션**을 만들고자 하는 것인지도 궁금합니다!

---

이번 미션에서도 많이 배워가도록 하겠습니다 :) 감사합니다 😆

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,199 @@
...
+
+### 역
+- [x] 역은 고유한 식별자를 가진다.
+- [x] 역 이름은 `역`으로 끝나야한다.
+- [x] 역 이름은 2글자에서 11글자까지 가능하다.
+- [x] 역 이름은 한글 + 숫자로만 이루어져야한다.
+
+### 구역
+- [x] 구역은 두 역과 역사이의 거리를 가진다.
+- [x] 거리는 0 이상의 정수이고, 최대 15이다. 단위는 km이다.
```
Comment (toneyparky) [inline README.md]: 이전 블랙잭에서도 리뷰 드렸듯 문서화는 현업에서 중요한 스킬 중 하나에요. (가끔은 코딩보다 문서를 더 많이 할 때도 있어요 ㅋㅋ)
게다가 READMD.md는 프로젝트의 얼굴과 같기에 더 중요한 문서죠. 

해당 문서에 바뀐 정책이 반영되어 있지 않다면 이를 보는 사람은 어떻게 생각할까요?

File: README.md [hunk 2]
```diff
@@ -1 +1,199 @@
...
+
+### 노선
+- [x] 노선 이름은 숫자 + `호선` 이다.
+  - [x] 숫자는 1 ~ 9까지 가능하다.
+
+- [x] 노선의 색은 `색`으로 끝나야한다.
+  - [x] 색 이름은 2글자에서 11글자까지 가능하다.
```
Comment (toneyparky) [inline README.md]: 임의로 요구사항을 제한해주셨네요! 덕분에 프로젝트의 로직이 명확해진거 같네요. 굳굳

File: README.md [hunk 3]
```diff
@@ -1 +1,199 @@
...
+
+- [x] 역을 제거할 수 있어야 한다.
+  -[x] 역을 제거할 경우 남은 역을 재배치 해야한다.
+  -[x] 노선에서 역이 제거될 경우 역과 역 사이의 거리도 재배정되어야한다.
+  -[x] 노선에 등록된 역이 2개 인 경우 하나의 역을 제거할 때 두 역이 모두 제거되어야한다.
+
+
+---
+
+## API 명세
```
Comment (toneyparky) [inline README.md]: API 명세를 추가해주셨네요..! 좋아요. 다음 미션에서는 프론트엔드와 협업을 하는 것으로 알고 있어요. 그 때에는 API 문서가 더 중요할거에요. 지금처럼 리드미에 관리해도 좋지만 API 스펙 변경시 매번 수동으로 수정해야하는 불편함이 있죠. 누락될 수도 있구요.

이를 해결하기 위해 Swagger, Rest Docs 라는 툴이 있어요. 두 툴을 비교하여 장단점을 정리 후 댓글로 남겨주세요. 그리고 다즐이 최종적으로 하나를 선택하여 적용해주시면 좋겠네요 😉

File: src/main/java/subway/controller/ControllerAdvice.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+@RestControllerAdvice
+public class ControllerAdvice {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ControllerAdvice.class);
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleException(final Exception exception) {
+        final String message = exception.getMessage();
+        LOGGER.warn(message);
```
Comment (toneyparky) [inline src/main/java/subway/controller/ControllerAdvice.java]: 로깅의 warn은 어떤 의미인가요? info, debug, error 등의 레벨이 있는데 warn으로 남긴 이유는 뭐에요?

File: src/main/java/subway/dao/StationDao.java [hunk 5]
```diff
@@ -1,58 +1,47 @@
...
 import org.springframework.jdbc.core.namedparam.SqlParameterSource;
 import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
 import org.springframework.stereotype.Repository;
-import subway.domain.Station;
-
-import javax.sql.DataSource;
-import java.util.List;
+import subway.entity.StationEntity;
 
 @Repository
```
Comment (toneyparky) [inline src/main/java/subway/dao/StationDao.java:12]: DAO와 Repository에 `@Repository` 어노테이션이 달려있군요. 

해당 어노테이션은 어떤 의미인지, 두 객체 모두에 적용하는 게 좋은지 아닌지에 대한 다즐의 생각도 적어주시면 감사하겠습니다 😎

File: src/main/java/subway/domain/line/Color.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        this.value = value;
+    }
+
+    private void validate(final String value) {
+        if (Objects.isNull(value)) {
+            throw new InvalidColorException("노선 색은 존재해야 합니다.");
+        }
+        if (!PATTERN.matcher(value).matches()) {
+            throw new InvalidColorException("색 이름은 '색'으로 끝나야 합니다.");
+        }
```
Comment (toneyparky) [inline src/main/java/subway/domain/line/Color.java]: 20글자지만 색으로 끝나는 Color를 생성하려하면 사용자는 어떤 예외를 확인하게 될까요?

File: src/test/java/subway/domain/line/ColorTest.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    @NullSource
+    void validateWithNull(final String input) {
+        assertThatThrownBy(() -> new Color(input))
+                .isInstanceOf(InvalidColorException.class)
+                .hasMessage("노선 색은 존재해야 합니다.");
+    }
+
+    @ParameterizedTest
+    @DisplayName("색 이름 형식에 맞지 않는다면 예외를 던진다.")
+    @ValueSource(strings = {"색", "하양", "white색", "하양하양하양하양하양하양색"})
```
Comment (toneyparky) [inline src/test/java/subway/domain/line/ColorTest.java]: `하양하양하양하양하양하양색` 이라는 테스트 케이스는 무얼 검증하고 싶은건지 한눈에 확인할 수 없네요. 

station의 `NameTest`에서 작성해주신 `열한글자가넘는역이름입니다역`처럼 수정해보면 더 좋겠네요~!

File: src/test/java/subway/domain/line/LineTest.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+            //when
+            //then
+            assertThatThrownBy(() -> line.addSection(newUpward, newDownward, 5))
+                    .isInstanceOf(InvalidSectionException.class)
+                    .hasMessage("연결할 역 정보가 없습니다.");
+        }
+
+        @Test
+        @DisplayName("역이 둘다 존재하지 않으면 예외를 던진다.")
```
Comment (toneyparky) [inline src/test/java/subway/domain/line/LineTest.java]: 해당 설명은 테스트의 내용과 매칭 될까요?

File: src/test/java/subway/domain/section/DistanceTest.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Test
+    @DisplayName("거리를 정상적으로 생성한다.")
+    void distance() {
+        assertDoesNotThrow(() -> new Distance(10));
+    }
+
+    @ParameterizedTest
+    @DisplayName("거리가 0보다 작을 경우 예외를 던진다.")
+    @ValueSource(ints = {-1, -3})
```
Comment (toneyparky) [inline src/test/java/subway/domain/section/DistanceTest.java:22]: -1과 -3을 테스트 케이스로 설정하신 이유가 궁금하네요~!

File: src/main/java/subway/exception/HttpException.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package subway.exception;
+
+import org.springframework.http.HttpStatus;
+
+public class HttpException extends RuntimeException {
```
Comment (toneyparky) [inline src/main/java/subway/exception/HttpException.java]: `HttpException`은 어떤 책임을 가진 객체에요?

File: src/main/resources/data.sql [hunk 11]
```diff
@@ -0,0 +1,27 @@
+CREATE TABLE IF NOT EXISTS STATION
```
Comment (toneyparky) [inline src/main/resources/data.sql:1]: 첫번째 질문에 대해 이야기 나눠보죠~ 

> DB에 있는 값은 안전하다고 판단할 수 있을까?

DB에 있는 값은 프로그램의 로직에 의해 저장이 되겠죠. 그렇기에 저희의 로직에 따라 안전할지 여부가 결정될 것 같아요. 로직에 버그가 있다면 안전하지 않은 값이 들어가겠죠. 그런 상황에서는 DB 의 값을 올바르게 맞춰줘야해요. 

코드를 짤 때에는 항상 안전한 값이 들어 있다고 가정하고 짜면 좋지 않을까요? 아니라면 매번 DB에서 받아온 값을 검증해줘야하니까요.
하지만 도메인 객체를 이용하면 자연스레 검증도 되겠네요. DB의 값을 도메인 객체로 만드는 과정에서 자연스레 유효성 검사가 될테니까요.

File: src/main/java/subway/service/LineService.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    private List<LineResponse> generateLineResponses(final List<Line> lines) {
+        return lines.stream()
+                .map(LineResponse::from)
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    @Transactional
+    public void createSection(final Long lineId, final SectionCreateRequest request) {
```
Comment (toneyparky) [inline src/main/java/subway/service/LineService.java:51]: 두번째 질문에 대해서도 이야기해보죠.

> 도메인 로직

말씀주신대로 각 방식은 장단점이 있다 생각합니다. 사람마다 기준이 다를 수 있는 부분이기도 하구요. 제 의견을 말씀드리자면 기본적으로는 도메인에 녹여내는게 좋다 생각해요. 우형도 예전에는 디비의 쿼리로 도메인 로직을 수행하던 적이 있다고 알고 있어요 ([관련 글](https://techblog.woowahan.com/2656/)) 아무래도 쿼리에 로직이 녹아들어있기에 테스트하기도 어렵고 요구사항이 바뀔 때에 대응하기도 어려웠을거에요 (저도 경험해보지는 못하여 추측입니다)

개발은 혼자하는 것이 아니기에 다즐이 적어주셨듯 유지보수의 목적이 클 것 같아요. 또한 코드를 이해하기도 쉽구요. SQL 쿼리 한줄과 명시적으로 표현된 함수 이름의 이해도는 다를테니까요..!

좋은 질문이네요 😁

**Review by toneyparky — CHANGES_REQUESTED** [review]
안녕하세요 다즐~! 오랜만이에요.

작성해주신 코드는 아주 깔끔하네요. 구조도 좋고 요구사항도 잘 반영되어 있구요. 특히 TDD를 적용하시며 꼼꼼하게 테스트해주신 부분이 인상적이에요 💯 

바로 머지할까도 했지만 조금 더 학습하면 좋을 부분이 있어 변경 요청 보내요. 
질문은 편하게 DM 주세요, 이번 미션도 화이팅입니다!

File: README.md (hunk 1 above)
Comment (woo-chang) [reply README.md]: 학습을 진행할 때도 문서를 먼저 보듯, 해당 프로젝트에 대해 이해하기 위해서는 문서를 가장 먼저 확인할 것이라 생각해요. 그런 문서에 바뀐 정책이 반영되어 있지 않다면 예상했던 동작이 발생하지 않아 혼란을 야기할 것이라 생각합니다. 🥲

개발은 혼자 진행하는 것이 아닌 협업을 통해 진행되기에 문서화를 잘 하는 것은 중요한 요소 중 하나라고 생각합니다. 그렇기에 문서화도 코드와 마찬가지로 유지보수 쉽게 작성하는 것도 중요할 것 같아요!

File: README.md (hunk 3 above)
Comment (woo-chang) [reply README.md]: > Swagger

Swagger는 어노테이션을 사용해서 API 문서를 자동으로 구성해주는 좋은 문서화 도구입니다. 자동으로 사용하기 좋은 인터페이스를 만들어주는 것도 큰 장점인 것 같아요. 하지만 프로덕션 코드에 문서화를 위한 코드가 들어간다는 단점도 존재합니다.

> Rest Docs

Rest Docs는 테스트 코드를 통해 API 문서가 작성되는 문서화 도구입니다. 테스트 코드를 통해 문서가 작성되기에 프로덕션 코드에 문서화를 위한 코드가 들어가지 않지만, Swagger처럼 예쁜 인터페이스가 적용되지 않고 사용법을 알아야 하기 때문에 진입점이 높은 것이 단점이라고 생각해요.

이번 미션에서는 쉽고 간단하게 API 문서를 만들어 낼 수 있는 `Swagger`를 적용해보고자 합니다!

File: src/main/java/subway/controller/ControllerAdvice.java (hunk 4 above)
Comment (woo-chang) [reply src/main/java/subway/controller/ControllerAdvice.java]: 로그의 레벨에는 5가지가 존재합니다.

- `error` : 예상하지 못한 심각한 문제가 발생하는 경우
- `warn` : 로직 상 유효성 확인과 같은 예상 가능한 문제로 인해 예외 처리
- `info` : 중요한 비즈니스 프로세스로 운영에 참고할만한 사항
- `debug` : 개발 단계에서 사용하는 로그
- `trace` : 모든 레벨에 대한 로그

Exception을 제외한 나머지 예외들은 저희가 처리하고자 하는 예외이기에 예상 가능한 문제로 인한 예외 처리에 속합니다. 따라서 지금 설정한 warn이 적절하다고 생각합니다. 하지만 예상 가능한 예외에서 잡히지 않고 Exception에 잡히는 경우 예상하지 못한 심각한 문제로 볼 수 있기에 `warn`이 아닌 `error`로 수정하는 것이 좋을 것 같아요 🙇🏻‍♂️

File: src/main/java/subway/dao/StationDao.java (hunk 5 above)
Comment (woo-chang) [reply src/main/java/subway/dao/StationDao.java:12]: `@Repository` 어노테이션은 우선 스프링 빈으로 등록하여 컴포넌트 스캔이 되도록 만들어줍니다. `@Component`와 다르게 데이터 접근 계층의 예외를 스프링 예외로 변환시켜주는 추가적인 기능을 제공하는 것으로 알고 있습니다.

Repository 객체도 도메인을 영속화시킨다는 관점에서 데이터 접근 계층에 속한다고 생각합니다. 그렇기에 명시적으로 `@Repository` 어노테이션을 통해 의미를 드러내는 것이 좋은 것 같습니다. 또한 추후에 `@Repository`에 추가적인 기능이 생기더라도 이를 활용할 수 있을 것 같아요. 해당 어노테이션과 Data JPA같은 기술이 함께 사용되면 프록시가 적용되어 이에 대한 비용이 발생할 수 있지만, 성능상에 문제가 되는 시점에 수정을 고려해볼 것 같습니다!

File: src/main/java/subway/domain/line/Color.java (hunk 6 above)
Comment (woo-chang) [reply src/main/java/subway/domain/line/Color.java]: 색으로 끝났는데 에러 원인을 확인할 수 없게 될 것 같습니다 🥲 예외를 조금 더 세분화해서 나누는 게 좋을 것 같아요!

File: src/test/java/subway/domain/line/ColorTest.java (hunk 7 above)
Comment (woo-chang) [reply src/test/java/subway/domain/line/ColorTest.java]: 실제 현업에서도 의미를 드러내기 위해 `열한글자가넘는역이름입니다역`을 사용하기도 하나요? ㅎㅎ

File: src/test/java/subway/domain/line/LineTest.java (hunk 8 above)
Comment (woo-chang) [reply src/test/java/subway/domain/line/LineTest.java]: 복사, 붙여넣기의 악영향이네요 .. 수정하겠습니다!

File: src/test/java/subway/domain/section/DistanceTest.java (hunk 9 above)
Comment (woo-chang) [reply src/test/java/subway/domain/section/DistanceTest.java:22]: 전체적으로 테스트 코드 작성은 `TDD`로 진행하였고, Edge Case인 -1만 해도 괜찮을 것 같지만, 코드가 -1에서만 예외가 발생하도록 잘못 작성되어 있을 경우도 고려하여 추가적인 경우도 테스트하려고 했습니다!

File: src/main/java/subway/exception/HttpException.java (hunk 10 above)
Comment (woo-chang) [reply src/main/java/subway/exception/HttpException.java]: 예외가 HttpStatus를 관리하도록 하기 위한 객체로 설계해보았습니다. 예외가 HttpStatus를 관리하도록 하여 예외에 대한 상태 코드 처리를 편하게 하려고 했으나 설명을 적다보니 어제의 저와 오늘의 저의 생각이 달라지게 된 것 같습니다.

HttpStatus는 컨트롤러 계층 즉, Presentation Layer에서 관리하고 알아야하는 정보입니다. 하지만 예외에서 이를 관리하게 되면 HttpStatus가 예외를 사용하는 곳인 Business Layer, Data Access Layer까지 내려갈 수 있게 되어버립니다. 이로 인해 하위 계층에서 상위 계층의 상태를 알게되어버리는 문제가 발생할 것 같아요.

그렇기에 예외에서는 메시지만 관리하고 HttpStatus는 컨트롤러 계층에 속해있는 ControllerAdvice에서 처리하도록 수정하는 것이 좋을 것 같아요. 이 부분에 대해서 토니의 생각도 들어보고 싶습니다!

File: src/main/resources/data.sql (hunk 11 above)
Comment (woo-chang) [reply src/main/resources/data.sql:1]: DB를 신뢰할 수 있도록 DB에 영향을 주는 로직은 더욱 꼼꼼히 테스트할 수 있도록 해야겠네요 👍 현업에서는 성능 이점을 얻기 위해 FK를 걸지 않는 경우도 있다고 들었는데 이런 경우 유효하지 않은 값이 들어갈 수 있는 가능성이 존재할 수도 있을 것 같아요. 이를 관리하기 위해서는 DB 상태를 지속적으로 확인하는 장치가 필요할 것 같은데 어떤 식으로 DB 상태를 유효하게 관리할 수 있는지 궁금합니다!

File: src/main/java/subway/service/LineService.java (hunk 12 above)
Comment (woo-chang) [reply src/main/java/subway/service/LineService.java:51]: 좋은 글과 좋은 얘기 공유해주셔서 감사합니다!

계속해서 의식적인 연습을 통해 각 상황에 알맞은 기술을 적용할 수 있도록 노력해야겠네요 🙇🏻‍♂️

File: src/main/java/subway/controller/ControllerAdvice.java (hunk 4 above)
Comment (woo-chang) [reply src/main/java/subway/controller/ControllerAdvice.java]: 한 가지 궁금한 점이 있습니다! `info`는 어떤 상황에 사용될지 감이 오지 않는 것 같아서 사용되는 예시를 알 수 있을까요? 🤔

File: README.md (hunk 1 above)
Comment (toneyparky) [reply README.md]: 네 문서를 만들면 유지보수를 잘 해야해요. 혹은 유지보수가 되지 않음을 명시하여 최신화된 정보를 따로 찾아야함을 알리는 방법도 있구요. 최고는 코드로 표현하는거지만 협업의 용이성을 위해 문서를 잘 관리해봅시다!

File: README.md (hunk 3 above)
Comment (toneyparky) [reply README.md]: 넵 추가해주신 거 확인했어요~! 각 툴의 장단점이 있기에 추후 프로젝트를 할 경우에 팀원들과 이야기를 나눠서 기술을 선택하면 좋겠네요.
고생하셨습니다!

File: src/main/java/subway/controller/ControllerAdvice.java (hunk 4 above)
Comment (toneyparky) [reply src/main/java/subway/controller/ControllerAdvice.java]: 굳굳 에러로 수정하는 데에 동의합니다. 서버 예외는 저희가 예측하지 못한 예외기에 빠르게 인식할 수 있어야하기 때문이죠. 

info는 말그대로 정보를 제공하는 로그 레벨이에요. 예를 들어 특정 시간마다 도는 애플리케이션이 동작을 시작할 때에 info 로그를 남겨서 개발자가 이를 인지할 수 있도록 할 수 있어요. 바로 확인해야 하거나 시스템에 문제가 발생한 상황이 아닐 때에 정보를 남기고프면 사용합니다.

File: src/main/java/subway/dao/StationDao.java (hunk 5 above)
Comment (toneyparky) [reply src/main/java/subway/dao/StationDao.java:12]: > Repository 객체도 도메인을 영속화시킨다는 관점에서 데이터 접근 계층에 속한다고 생각합니다.

다즐의 관점을 적어주셔서 감사해요. 이해되었습니다~! 굳굳

File: src/main/java/subway/domain/line/Color.java (hunk 6 above)
Comment (toneyparky) [reply src/main/java/subway/domain/line/Color.java]: 네넵 예외 메시지를 명확하게 표현하는게 중요하기에 유의해주시면 좋겠어요. 특히 도메인 레이어는 더 철저해야하죠. 수정사항 확인했습니다.

File: src/test/java/subway/domain/line/ColorTest.java (hunk 7 above)
Comment (toneyparky) [reply src/test/java/subway/domain/line/ColorTest.java]: ㅋㅋㅋ 사용할 수 있죠. 테스트 코드를 이해하기 쉽도록 하는 데에 목적이 있으니까요.

File: src/test/java/subway/domain/section/DistanceTest.java (hunk 9 above)
Comment (toneyparky) [reply src/test/java/subway/domain/section/DistanceTest.java:22]: 네넵 -1만 검증하기에는 구멍이 있다고 느끼셨군요. `Integer.MIN_VALUE` 같은 값을 이용하면 음수에 대해 테스트한다는 내용을 보다 명확하게 표현할 수 있을 것 같네요. 지금도 적어주신 이유가 이해되어 따로 수정하실 필요는 없겠어요~!

File: src/main/java/subway/exception/HttpException.java (hunk 10 above)
Comment (toneyparky) [reply src/main/java/subway/exception/HttpException.java]: 크.. 좋은 사고의 흐름이네요. 각 레이어의 책임에 대해서 명확하게 인지하고 계시다 생각해요. 저도 다즐의 의견에 동의합니다. HTTP라는 용어 자체가 컨트롤러 레이어가 아닌 다른 레이어에서 사용된다는 점만으로도 이상함을 느낄 수 있어요. (물론 HTTP가 도메인 개념에 들어있거나 유즈케이스에 HTTP라는 개념이 포함된다면 도메인 레이어와 서비스 레이어에 들어갈 수 있겠지만, 일반적인 상황을 말씀드려요~!)

잘 수정해주셨다 생각합니다~

File: src/main/resources/data.sql (hunk 11 above)
Comment (toneyparky) [reply src/main/resources/data.sql:1]: 맞아요 FK를 걸지 않는 경우도 있어요. 이 경우에서는 말씀주신 것 처럼 하나의 테이블에서 값을 제거하고 약하게 연결된 다른 테이블에서는 제거하지 않는다면 신뢰할 수 없는 데이터가 되겠죠. 
요 부분도 따로 상태를 관리하는 장치가 있는지는 모르겠어요. 다만 DB에 직접 접근하여 삭제함을 지양하고 삭제도 로직을 잘 작성하여 버그 없이 삭제가 진행될 수 있도록 처리해야해요. 마치 생성시에도 신경을 쓰듯 삭제, 수정시에도 신경을 쓰고 이를 코드로 잘 관리하고 테스트도 한다고 이해해주시면 될 것 같아요. 

만일 이미 DB의 상태가 깨진 상황이라면, DB 데이터를 전수 조사하지 않을까해요. 물론 이도 1회성 배치 애플리케이션 코드를 작성하여 해결할 것 같네요.

File: src/main/java/subway/service/LineService.java (hunk 12 above)
Comment (toneyparky) [reply src/main/java/subway/service/LineService.java:51]: 맞아요. 쿼리로 작성하면 뚝딱 처리할 수 있을 것만 같지만 그만큼 유지보수하기에 어려울 수 있어요. 레벨 3 프로젝트를 진행하시면서 관련된 고민을 하실텐데 의식적으로 연습해보시길..!

**Review by toneyparky — APPROVED** [review]
주말에도 고생하셨습니다 다즐~!
1단계는 머지할게요. 2단계 가시죠~~!
